File: Python/n-ary-tree-preorder-traversal.py
from typing import List


# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class Solution:
    def preorder(self, root: Node) -> List[int]:
        # Time:  O(n)
        # Space: O(n)
        result = []
        if root:
            self._preorder(root, result)
        return result

    def _preorder(self, node: Node, result: List[int]) -> None:
        result.append(node.val)
        for child in node.children:
            self._preorder(child, result)


class Solution2:

    def preorder(self, root: Node) -> List[int]:
        # Time:  O(n)
        # Space: O(n)
        if not root:
            return []

        result = []
        stack = [root]
        while stack:
            node = stack.pop()
            result.append(node.val)
            stack.extend(reversed(node.children))
        return result

File: C++/n-ary-tree-preorder-traversal.cpp
#include <vector>


// Definition for a Node.
class Node {
public:
    int val;
    std::vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, std::vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};


class Solution {
public:
    std::vector<int> preorder(Node* root) {
        // Time:  O(n)
        // Space: O(n)
        std::vector<int> result;
        if (root) {
            _preorder(root, result);
        }
        return result;
    }
private:
    void _preorder(Node* node, std::vector<int>& result) {
        result.push_back(node->val);
        for (auto child: node->children) {
            _preorder(child, result);
        }
    }
};